keps/sig-node/5328-node-capabilities/README.md

```diff
@@ -0,0 +1,987 @@
+    *   If a feature is disabled or unsupported, the Kubelet must remove the corresponding capability key from the map.
+    *   This approach ensures consistency and simplifies logic for the control plane, treating nodes that don't know about a feature the same as those that have it disabled.
+
+4.  Validation Rules
+    *   Keys and values within the `node.status.capabilities` map must adhere to the same validation rules as standard Kubernetes labels.
+    *   The Kubelet validates each capability and will discard any invalid key-value pair, logging an error.
+    *   When introducing a new capability that requires a Kubelet change, developers must ensure that the new keys and values meet these validation rules. This would also be enforced by [unit tests](#unit-tests).
+
+
+**Example:**
+
+```
+capabilities:
+   compatibility.kubernetes.io/guaranteed-qos-pod-cpu-resize: "true"
+```
+
+### Kubelet Changes
+
+Kubelet has the primary responsibility of discovering the capabilities during bootstrap and updating `node.status.capabilities`. It determines its full, static set of capabilities **once** upon startup based on its configuration. This list is held in memory and is included in every periodic `NodeStatus` update that the Kubelet sends to the API server. The Kubelet is the authoritative source of truth for its own capabilities. If an external controller or user were to manually edit the `node.status.capabilities` field on the Node object, the Kubelet's next periodic status update would automatically overwrite and revert those changes.
+
+In addition to reporting, Kubelet will also need to validate the pod. Before admitting a pod to be run on the node, the Kubelet's pod admission handlers check if the pod requires any specific capabilities (as inferred from its spec). It will validate these requirements against its own live, in-memory map of supported capabilities. If a required capability is not present, the Kubelet will reject the pod, transition it to a Failed phase, and post an appropriate event to the API server detailing the missing capability. This secondary validation (along with kube-scheduler capability based filtering) is necessary to handle node restart scenarios where a capability that existed during pod scheduling does not exist anymore (feature gate flip with node restart).
+
+As part of the feature's graduation to GA and eventual removal of the feature gate, the Kubelet must stop reporting the capability. This ensures that obsolete information is removed from the `node.status` object as part of the standard feature cleanup process.
+
+### Shared Capability Matching Library
+
+To avoid code duplication and ensure that all components make decisions based on the same logic, a new shared library will be introduced in the `k8s.io/component-helpers/nodecapabilities` staging repository. This library encapsulates all the logic for inferring and matching pod capability requirements.
+
+**Capability Registration**
```

Comment:
(already discussed offline out-of-band, adding notes here)

@liggitt pointed out that the library should be consumable by controllers that interact with multiple Kubernetes clusters that aren't necessarily the same version, which means that the library should be able to support multiple kubernetes versions.

I think the best way to handle this is to associate a Min & Max version with each capability (provided in registration), and include target version as an input when inferring the requirements.

Reply:
Updated. `MinVersion` and `MaxVersion` are passed during registration and `targetVersion` is used while checking pod's feature requirements.

Reply:
Following up on this, I'm pretty sure we can drop the MinVersion, and in fact that might be desireable. The _library_ should have a notion of min & max version, so if you ask it to derive the required features outside that window it can throw an error, but for a given pod, I think it's better to state that it requires an unsupported feature rather than silently dropping it.